[ResourceMonitor] Add more meaningful release log information part 1.

basuke opened the pull request:
#### b47957d7891a2b461cf18eb5e627f8e368626905
<pre>
[ResourceMonitor] Add more meaningful release log information part 1.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288171">https://bugs.webkit.org/show_bug.cgi?id=288171</a>
<a href="https://rdar.apple.com/145259890">rdar://145259890</a>

Reviewed by Per Arne Vollan and Chris Dumez.

Adding new log category and move current release logs to that category. And also add more meaningful information
for the record.

This is part 1, targeting WebCore.

In this change, chances are:
- URL checks are most of two times per each iframe.
- One check generates several logs.
- When it hits the network limits, some extra logs is produced.
- If unloading happens, ~10 more logs.

One exception is for shared worker, which logs all network usage and this will be rare case. In case it causes an issue,
this will be removed. For now, it is required to observe the usage during development.

* LayoutTests/http/tests/iframe-monitor/dark-mode-expected.txt:
* LayoutTests/http/tests/iframe-monitor/eligibility-expected.txt:
* LayoutTests/http/tests/iframe-monitor/iframe-unload-expected.txt:
* LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt:
* LayoutTests/http/tests/iframe-monitor/throttler-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitor):
* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::resourceTypeToString):
* Source/WebCore/loader/ResourceLoadInfo.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::didReceiveResponse):
(WebCore::eligibilityToString):
(WebCore::ResourceMonitor::continueAfterDidReceiveEligibility):
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::checkEligibility):
* Source/WebCore/loader/ResourceMonitorPersistence.cpp:
(WebCore::ResourceMonitorPersistence::reportSQLError):
(WebCore::ResourceMonitorPersistence::openDatabase):
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::~ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):
(WebCore::LocalFrame::reportResourceMonitoringWarning):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::reportNetworkUsage):

Canonical link: <a href="https://commits.webkit.org/290850@main">https://commits.webkit.org/290850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f177bb3b04dedce9ae9c8ed0fd37bf50dd8b0b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10822 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/317 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96268 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42007 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93337 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19139 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/96268 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27630 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82689 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/96268 "") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/270 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41150 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/280 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98260 "") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18460 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/98260 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18716 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78524 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/98260 "") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19371 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22844 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18459 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18178 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21639 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/19950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -41,7 +41,8 @@ namespace WebCore {
 #if ENABLE(CONTENT_EXTENSIONS)
 
-#define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) RELEASE_LOG(ResourceLoading, "%p - ResourceMonitor(frame %p)::" fmt, this, m_frame.get(), ##__VA_ARGS__)
+#define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) RELEASE_LOG(ResourceMonitoring, "%p - ResourceMonitor(frame %p)::" fmt, this, m_frame.get(), ##__VA_ARGS__)
```
Would it be OK to log the frame ID instead of the frame pointer? Also, is it required to log the pointer of the ResourceMonitor? This will increase the log volume, and could be good to avoid, if possible.

basuke replied:
Thanks! I'll fix that.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -85,6 +87,11 @@ void ResourceMonitor::setDocumentURL(URL&& url)
 
+static ASCIILiteral eligibilityToString(ResourceMonitorEligibility eligibility)
+{
+    return (eligibility == ResourceMonitorEligibility::Eligible ? "eligible"_s : "not eligible"_s);
```
Parentheses are not needed.

cdumez commented on Source/WebCore/loader/ResourceLoadInfo.cpp:
```diff
@@ -159,6 +159,37 @@ ResourceFlags ResourceLoadInfo::getResourceFlags() const
     return flags;
 }
 
+ASCIILiteral resourceTypeToString(OptionSet<ResourceType> resourceType)
```
This looks very inefficient, I suspect we can iterate over the OptionSet, do a switch statement on the first value and return right away.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -104,9 +110,22 @@ void ResourceMonitor::didReceiveResponse(const URL& url, OptionSet<ContentExtens
-    ResourceMonitorChecker::singleton().checkEligibility(WTFMove(info), [weakThis = WeakPtr { *this }](Eligibility eligibility) {
-        if (RefPtr protectedThis = weakThis.get())
+    ResourceMonitorChecker::singleton().checkEligibility(WTFMove(info), [this, weakThis = WeakPtr { *this }, url, resourceType](Eligibility eligibility) {
+        if (RefPtr protectedThis = weakThis.get()) {
```
Please do an early return instead.

pvollan commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -61,9 +61,10 @@ void ResourceMonitor::setEligibility(Eligibility eligibility)
-    RESOURCEMONITOR_RELEASE_LOG("The frame is %" PUBLIC_LOG_STRING ".", (eligibility == Eligibility::Eligible ? "eligible" : "not eligible"));
 
     if (isEligible()) {
+        RESOURCEMONITOR_RELEASE_LOG("Frame (%" SENSITIVE_LOG_STRING ") was set as eligible.", m_frameURL.utf8().data());
```
For logs that are emitted from the WebContent process, you may also consider using `RELEASE_LOG_FORWARDABLE`, which is more efficient.

pvollan commented on Source/WebCore/workers/shared/SharedWorker.cpp:
```diff
@@ -200,8 +200,10 @@ void SharedWorker::reportNetworkUsage(size_t bytesTransferredOverNetwork)
     if (delta) {
-        if (RefPtr resourceMonitor = m_resourceMonitor)
+        if (RefPtr resourceMonitor = m_resourceMonitor) {
+            RELEASE_LOG(ResourceMonitoring, "[identifier=%" PUBLIC_LOG_STRING "] SharedWorker::reportNetworkUsage to ResourceMonitor: %zu bytes", identifier().toString().utf8().data(), delta.value());
```
Would it be more efficient to log this as an identifier, instead of converting to a string?

pvollan reviewed (approved):
R=me.

basuke committed:
[ResourceMonitor] Add more meaningful release log information part 1.
https://bugs.webkit.org/show_bug.cgi?id=288171
rdar://145259890

Reviewed by Per Arne Vollan and Chris Dumez.

Adding new log category and move current release logs to that category. And also add more meaningful information
for the record.

This is part 1, targeting WebCore.

In this change, chances are:
- URL checks are most of two times per each iframe.
- One check generates several logs.
- When it hits the network limits, some extra logs is produced.
- If unloading happens, ~10 more logs.

One exception is for shared worker, which logs all network usage and this will be rare case. In case it causes an issue,
this will be removed. For now, it is required to observe the usage during development.

* LayoutTests/http/tests/iframe-monitor/dark-mode-expected.txt:
* LayoutTests/http/tests/iframe-monitor/eligibility-expected.txt:
* LayoutTests/http/tests/iframe-monitor/iframe-unload-expected.txt:
* LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt:
* LayoutTests/http/tests/iframe-monitor/throttler-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitor):
* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::resourceTypeToString):
* Source/WebCore/loader/ResourceLoadInfo.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::didReceiveResponse):
(WebCore::eligibilityToString):
(WebCore::ResourceMonitor::continueAfterDidReceiveEligibility):
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::checkEligibility):
* Source/WebCore/loader/ResourceMonitorPersistence.cpp:
(WebCore::ResourceMonitorPersistence::reportSQLError):
(WebCore::ResourceMonitorPersistence::openDatabase):
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::~ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):
(WebCore::LocalFrame::reportResourceMonitoringWarning):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::reportNetworkUsage):

Canonical link: https://commits.webkit.org/290850@main
